packages/async-rewriter2/src/async-writer-babel.ts

```diff
@@ -0,0 +1,575 @@
+        // Finally, do not wrap constructor functions. This is not a technical
+        // necessity, but rather a result of the fact that we can't handle
+        // asynchronicity in constructors well (e.g.: What happens when you
+        // subclass a class with a constructor that returns asynchronously?).
+        if (path.parentPath.isClassMethod() &&
+            path.parentPath.node.key.type === 'Identifier' &&
+            path.parentPath.node.key.name === 'constructor') {
+          return;
+        }
+
+        // A parent function might have a set of existing helper methods.
+        // If it does, we re-use the functionally equivalent ones.
+        const existingIdentifiers =
+          path.findParent(path => !!path.getData(identifierGroupKey))?.getData(identifierGroupKey);
+
+        // Generate and store a set of identifiers for helpers.
+        const functionState = path.scope.generateUidIdentifier('fs');
+        const synchronousReturnValue = path.scope.generateUidIdentifier('srv');
+        const asynchronousReturnValue = path.scope.generateUidIdentifier('arv');
+        const expressionHolder = path.scope.generateUidIdentifier('ex');
+        const markSyntheticPromise = existingIdentifiers?.markSyntheticPromise ?? path.scope.generateUidIdentifier('msp');
+        const isSyntheticPromise = existingIdentifiers?.isSyntheticPromise ?? path.scope.generateUidIdentifier('isp');
+        const syntheticPromiseSymbol = existingIdentifiers?.syntheticPromiseSymbol ?? path.scope.generateUidIdentifier('sp');
+        path.parentPath.setData(identifierGroupKey, {
+          functionState,
+          synchronousReturnValue,
+          asynchronousReturnValue,
+          expressionHolder,
+          markSyntheticPromise,
+          isSyntheticPromise,
+          syntheticPromiseSymbol
+        });
+
+        // We generate code that vaguely looks like and insert it at the top
+        // of the wrapper function:
+        // const syntheticPromise = Symbol.for('@mongosh.syntheticPromise');
+        // function markSyntheticPromise(promise) {
+        //   return Object.defineProperty(promise, syntheticPromise, { value: true });
+        // }
+        // function isSyntheticPromise(value) {
+        //   return value && value[syntheticPromise];
+        // }
+        // Note that the last check potentially triggers getters and Proxy methods
+        // and we may want to replace it by something a bit more sophisticated.
+        // All of the top-level AST nodes here are marked as generated helpers.
+        const runtimeSupport = existingIdentifiers ? [] : [
+          Object.assign(t.variableDeclaration('const', [
+            t.variableDeclarator(
+              syntheticPromiseSymbol,
+              t.callExpression(
+                t.memberExpression(symbolConstructor, t.identifier('for')),
+                [t.stringLiteral('@@mongosh.syntheticPromise')]
+              )
+            )
+          ]), { [isGeneratedHelper]: true }),
```

Comment:
Maybe we could replace those code generation parts with [@babel/template](https://babeljs.io/docs/en/babel-template) so that it's more readable at a glance, it requires quite some effort to process what those ast types represent

Reply:
Yeah, so … I’m still a babel newbie, and I wasn’t quite sure how to integrate that with the fact that identifiers are dynamically generated (to avoid conflicts), but I agree, that would make things a bit more readable :)

Reply:
Yeah that's something I also had a hard time digesting and understanding :D The templates thing looks awesome for that :) Maybe we can do that together @gribnoysup and do some pairing?

Reply:
Not a babel pro too, just suggesting 😆 Yeah, with the amount of stuff we dynamically replace it might be not that easy and readable actually, but I'd give it a shot if you have no objections @addaleax

@rose-m totes!

Reply:
Oh, yeah, like I said, feel free to make as many changes as you'd like here 👍